### PR TITLE
 authenticate clients based on certificate CommonName in v3 API

### PR DESCRIPTION
**Pull request by mitake**
This PR lets v3 auth mechanism authenticate clients based on CommonName of certificate like v2 auth.

/cc @yudai

### CODE REVIEW

**Comment by xiang90** [comment]
@mitake Can you resolve the conflict? Also we need to add documentation around it.

**Comment by mitake** [comment]
@xiang90 sure, I'll resume this PR and the jwt token support from next week.

**Comment by yudai** [comment]
@mitake 
I ran some tests and confirmed:

* CommonName in client certs are read properly
* values given to `--user` do not overwrite CommonName in client certs
    * but a token is generated internally, i'm not sure if it's safe
* client certs signed by an invalid CA (not the one given to `--trusted-ca-file`) are rejected
    * btw, in this case, client gets `Error:  grpc: timed out when dialing` which is a bit confusing

It looks good overall.

**Comment by mitake** [comment]
updated for e2e testing issue.

@yudai thanks for testing, I'll consider about the behaviors you pointed out tomorrow.

File: auth/store.go [hunk 1]
```diff
@@ -879,7 +883,7 @@ func (as *authStore) isAuthEnabled() bool {
 	return as.enabled
 }
 
-func NewAuthStore(be backend.Backend, indexWaiter func(uint64) <-chan struct{}) *authStore {
+func NewAuthStore(be backend.Backend, indexWaiter func(uint64) <-chan struct{}, clientCertAuthEnabled bool) *authStore {
```
Comment (xiang90) [inline auth/store.go]: i dislike boolean arg in a public api.

@heyitsanthony opinions?

File: auth/store.go [hunk 2]
```diff
@@ -950,7 +955,30 @@ func (as *authStore) isValidSimpleToken(token string, ctx context.Context) bool
 	return false
 }
 
+func (as *authStore) authInfoFromPeer(tlsInfo credentials.TLSInfo) (*AuthInfo, error) {
```
Comment (xiang90) [inline auth/store.go]: authInfoFromTLS?

File: auth/store.go (hunk 1 above)
Comment (heyitsanthony) [inline auth/store.go]: Is there any reason why should be optional? Could just hardwire it to always be enabled.

File: e2e/etcd_test.go [hunk 3]
```diff
@@ -106,6 +106,13 @@ var (
 		isPeerTLS:    true,
 		initialToken: "new",
 	}
+	configClientTLSCertAuth = etcdProcessClusterConfig{
+		clusterSize:           3,
```
Comment (xiang90) [inline e2e/etcd_test.go]: do we really need a 3 member cluster for this test?

Comment (mitake) [inline e2e/etcd_test.go]: Yes, 3 nodes are needless. I'll reduce the number of nodes.

File: auth/store.go (hunk 1 above)
Comment (mitake) [inline auth/store.go]: Yes, on the second thought, the condition based on `Cfg.ClientCertAuthEnabled` seems to be not so meaningful. I'd like to remove it. Maybe making a clear rule about priorities of auth methods (a token generated by `Authenticate()` and common name of cert) and documenting it is required.

Comment (mitake) [inline auth/store.go]: On the third thought, the condition would be needed for not using common names in auth accidentally (the option is provided for this purpose in v2 auth). How about passing `Cfg.ClientCertAuthEnabled` via `ctx` of `AuthInfoFromCtx()`?

**Comment by mitake** [comment]
@xiang90 @heyitsanthony updated for your comments, could you take a look? Now I separated `auth.AuthInfoFromTLS()` from `auth.AuthInfoFromCtx()` and let etcdserver call them based on its configuration for avoiding complicated initialization of AuthStore.

**Comment by xiang90** [comment]
LGTM

File: etcdserver/v3_server.go [hunk 4]
```diff
@@ -802,3 +802,14 @@ func (s *EtcdServer) linearizableReadNotify(ctx context.Context) error {
 		return ErrStopped
 	}
 }
+
+func (s *EtcdServer) AuthInfoFromCtx(ctx context.Context) (*auth.AuthInfo, error) {
+	if s.Cfg.ClientCertAuthEnabled {
```
Comment (heyitsanthony) [inline etcdserver/v3_server.go]: @xiang90 is there any reason to keep this as optional in v3 if it'll just fall back to regular user auth?

Comment (xiang90) [inline etcdserver/v3_server.go]: admin can disable cert auth. If we do not check here we might make mistake, like now command name "foo" == user "foo".

**Comment by heyitsanthony** [comment]
lgtm thanks

**Comment by xiang90** [comment]
@mitake Merging this for you! Thanks a lot!

**Comment by mitake** [comment]
@xiang90 @heyitsanthony thanks!

@yudai I'll work on the points you mentioned later because they would be independent from this PR (they are related to client and transport, I think)

**Comment by jessfs** [comment]
@mitake it doesn't look like the documentation ever got completed for how to use this feature. I've checked all the related issues as well such as #7251 and #3916. I did find a comment from you [here](https://github.com/coreos/etcd/pull/3916#issuecomment-283359359) mentioning you were going to work on it soon, but that's where the trail seems to go cold.

**Pull request merged**